### PR TITLE
8300979: Lazily initialize (byte, char)arr in java.io.DataInputStream

### PR DESCRIPTION
**Pull request by minborg**
This PR proposed to lazily initialize the scratch arrays only if/when needed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8300979](https://bugs.openjdk.org/browse/JDK-8300979): Lazily initialize (byte, char)arr in java.io.DataInputStream

### Reviewers
 * @eirbjo (no known openjdk.org user name / role) ⚠️ Review applies to [9d99127b](https://git.openjdk.org/jdk/pull/13121/files/9d99127b11430c3761afe796597153479a990ba8)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13121/head:pull/13121` \
`$ git checkout pull/13121`

Update a local copy of the PR: \
`$ git checkout pull/13121` \
`$ git pull https://git.openjdk.org/jdk.git pull/13121/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13121`

View PR using the GUI difftool: \
`$ git pr show -t 13121`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13121.diff">https://git.openjdk.org/jdk/pull/13121.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/io/DataInputStream.java [hunk 1]
```diff
@@ -573,7 +573,7 @@ public static final String readUTF(DataInput in) throws IOException {
         byte[] bytearr = null;
         char[] chararr = null;
         if (in instanceof DataInputStream dis) {
-            if (dis.bytearr.length < utflen){
+            if (dis.bytearr == null || dis.bytearr.length < utflen){
```
Comment (eirbjo) [inline src/java.base/share/classes/java/io/DataInputStream.java:576]: Add a space between ')' and '{'?

```suggestion
            if (dis.bytearr == null || dis.bytearr.length < utflen) {
```

**Review by eirbjo — APPROVED** [review]
Looks good. Many DataInputStreams probably never read UTF. (Not a reviewer)

File: src/java.base/share/classes/java/io/DataInputStream.java [hunk 2]
```diff
@@ -61,8 +61,8 @@ public DataInputStream(InputStream in) {
     /**
      * working arrays initialized on demand by readUTF
      */
-    private byte[] bytearr = new byte[80];
-    private char[] chararr = new char[80];
+    private byte[] bytearr;
```
Comment (bplb) [inline src/java.base/share/classes/java/io/DataInputStream.java:64]: Could the `bytearr` and `chararr` instance variables be removed altogether?

**Comment by eirbjo** [comment]
Looks like `DataInputStream.readUTF` could benefit from using `JavaLangAccess.countPositives` and return early for the common case that all bytes are ASCII. Food for another PR?

**Comment by stsypanov** [comment]
Nice idea. Just wonder could we do the same for `BufferedInputStream`?

**Comment by minborg** [comment]
> Looks like `DataInputStream.readUTF` could benefit from using `JavaLangAccess.countPositives` and return early for the common case that all bytes are ASCII. Food for another PR?

Yep. Please raise an issue with the ideas around this if deemed beneficial.

**Comment by minborg** [comment]
> Nice idea. Just wonder could we do the same for `BufferedInputStream`?

I think it is doable but would require some more work as nullity is used as a "closed flag". Please raise a separate issue about this.

**Review by AlanBateman — APPROVED** [review]

**Comment by eirbjo** [comment]
> The reason this puzzles me is that the `readInt` benchmark don't involve the UTF code paths at all. So why would it become slower, just because two fields are no longer initialized with the default arrays?

Could there be some JIT optimizations not being possible because the fields are no longer 'never null'?

**Comment by altrisi** [comment]
If that's the case an alternative idea could be to make them hold a (shared) empty array by default instead, that'd also match the condition for entering the branch and wouldn't be null.

**Comment by minborg** [comment]
@eirbjo I've updated the PR so we hold zero-length arrays rather than null arrays. Please re-run your benchmark to see if there is any change.

**Comment by minborg** [comment]
With the new non-null proposal and on my Mac M1, I get:

Baseline:
```
Benchmark                     Mode  Cnt  Score   Error  Units
DataInputStreamTest.readChar  avgt   20  5.077 ± 0.052  us/op
DataInputStreamTest.readInt   avgt   20  2.630 ± 0.014  us/op
```

PR (updated version):
```
Benchmark                     Mode  Cnt  Score   Error  Units
DataInputStreamTest.readChar  avgt   20  5.052 ± 0.028  us/op
DataInputStreamTest.readInt   avgt   20  2.615 ± 0.022  us/op
```

So, on my platform, the performance is unchanged.

**Review by AlanBateman — APPROVED** [review]

**Comment by eirbjo** [comment]
> @eirbjo I've updated the PR so we hold zero-length arrays rather than null arrays. Please re-run your benchmark to see if there is any change.

@minborg I ran the test now after your integration and I no longer observe the regression I did before.

**Comment by minborg** [comment]
> > @eirbjo I've updated the PR so we hold zero-length arrays rather than null arrays. Please re-run your benchmark to see if there is any change.
> 
> @minborg I ran the test now after your integration and I no longer observe the regression I did before.

Thanks @eirbjo for finding and verifying the performance regression.

**Pull request closed without merging**